README.md
# EC-Admin

EC-Admin is a collection of advanced management, optimization, and customization
tools for CloudBees Flow. It includes tools for:

* Bulk resource health checks
* White labeling
* Log management
* Advanced backup
* License statistics
* Debug tools
* More

EC-Admin is used in many CloudBees Flow field deployments.

In June 2018, the plugin was converted to the [Plugin Wizard DSL framework](https://github.com/electric-cloud/PluginWizard)
format so the whole plugin code is now managed on GitHub without the need
to extract the project first. In March 2019, the plugin was furthermore
converted to the new [plugin builder](https://github.com/electric-cloud/ec-plugin-tool)
to inprove promotion time

The official stable version can be found in the plugins catalog or in
our [plugins directory](https://electric-cloud.com/plugins/directory/p/ec-admin).
You can also follow development in our [Community GitHub repository](https://github.com/electric-cloud-community/EC-Admin).
For the latest code changes, check the [maintainer's fork](https://github.com/lrochette/EC-Admin)

Requirements:

* CloudBees Flow 6.0 minimum
* EC-PluginManager 1.4.0


## Installation

It is available as a plugin delivered in the form of a .jar file.

To install the plugin, use one the following methods:

* CloudBees Flow Administration/Plugins UI - "Install from File/URL"
* CLI - `ectool installPlugin --force 1 EC-Admin.jar`
* Perl - `$ec->installPlugin()` API.

Don't forget to promote the plugin after installation.

To install on 4.0.x or or 4.1.x, import the EC_4.0.xml file. Be
aware that some features are not present as the original project
makes use of `createJobStep()` API which was introduced in 4.2.0:

```ectool import /path_to/EC-Admin_for_EC_4.0.xml --disableSchedules 1```



# Contact authors

* License Logger:
  * Mike Westerhof

* deleteObjects:
* testResources:
  * [Tanay Nagjee](https://github.com/tanaynagjee)

* Everything else:
  *  [Laurent Rochette](mailto:lrochette@electric-cloud.com)


# Thanks
Thanks to Mark Hall for providing the Dev and Prod logos for changeBannerColor.

Thanks to [David Hubbell](mailto:dhubbell@spkaa.com) for his constant flow of
fixes to my poor code.

# Legal Mumbo Jumbo

This plugin is free for use. Modify it however you see fit to better your
experience using CloudBees Flow. Share
your [enhancements](https://github.com/electric-cloud-community/EC-DslDeploy/issues)
and [fixes](https://github.com/electric-cloud-community/EC-DslDeploy/pulls)

This module is not officially supported by Electric Cloud. It has undergone no
formal testing and you may run into issues that have not been uncovered in the
limited testing done so far.

[Electric Cloud](https://www.electric-cloud.com) should not be held liable for
any repercussions of using this plugin.


CHANGELOG
2019-06-05 Laurent Rochette <lrochette@cloudbees.com>
  * 4.3.0: issue 38: add support for devOpsInsightDataSources in saveAllObjects

2019-04-30 Laurent Rochette <lrochette@cloudbees.com>
  * 4.2.1: remove generateDslTree
           Fix cleanup spec test
  * 4.2.0: Replace default logo with new CloudBees Flow one
           Rebrand ElectricFlow into CloudBees Flow

2019-04-19 Laurent Rochette <lrochette@electric-cloud.com>
  * 4.1.1: Issue with saveAllObjects on catalogItems

2019-04-17 Laurent Rochette <lrochette@electric-cloud.com>
  * 4.1.0: Save email configurations and directory providers

2019-03-28 Laurent Rochette <lrochette@electric-cloud.com>
  * 4.0.0: Fix #31 Convert to new plugin builder to speed up promotion

2019-03-20 Laurent Rochette <lrochette@electric-cloud.com>
  * 3.6.1: Fix #5 (formerly #119) New procedure to clean old pipelineRuntimes
           Help page modified

2019-03-12 Laurent Rochette <lrochette@electric-cloud.com>
  * 3.5.1.814: Stop agentMemnoryConfiguration when data is incorrect
               Convert tests to ec-spec-tools

2019-02-25 Laurent Rochette <lrochette@electric-cloud.com>
  * 3.5.0.813: Fix #165: backup dashboards, widgets, reports with saveAllObjects
               backup schedules with saveAllObjects
               Fix #121: improve tooltip for "olderThan" in jobsCleanup
2019-02-22 Laurent Rochette <lrochette@electric-cloud.com>
  * 3.4.2.800: Fix #168: Ping agent when changing memory configuration
  * 3.4.1.792: Fix non-deletion of the jobs when the schedule is in a different
      project for licensing
  * 3.4.0.791: Fix #166 - backup catalogs with saveAllObjects
  * 3.3.1.786: Changing paths in saveAllObjects to match EC-DslDeploy

2019-02-11 Laurent Rochette <lrochette@electric-cloud.com>
  * 3.2.7.775: Fix #163 merge saveAllObjects and saveDslObjects
  * 3.2.6.769: Reactivate agentMemoryConfiguration and fixed #168.

2019-02-10 Laurent Rochette <lrochette@electric-cloud.com>
  * 3.2.5.763: Add save tags (issue #161)
               Add save services (issue #161)
               Add save personas, personaPages and personaCategories (issue #161)
               saveDSL directories are in lower case to match EC-DslDeploy

2019-02-08 Laurent Rochette <lrochette@electric-cloud.com>
  * 3.2.4.748 : Add generateDsl procedure
                Fix #162: Add exportPlugins

2019-01-23 Laurent Rochette <lrochette@electric-cloud.com>
  * 3.2.3.736: Fix #156 issue with getP since 8.5

2019-01-04 Laurent Rochette <lrochette@electric-cloud.com>
  * 3.2.2.733: Fix #158: max the number of job deleted in jobsCleanup (NMB-24602 and NMB-25946)

2018-12-09 Laurent Rochette <lrochette@electric-cloud.com>
  * 3.2.1.728: Fix #156: references to src are not valid anymore

2018-11-13 David Hubbell <dhubbell@spkaa.com>
  * 3.2.1.726: fix #154: jobsCleanup fails when trying to expand windows path

2018-11-13 Laurent Rochette <lrochette@electric-cloud.com>
  * 3.2.0.725: Remove old and unused procedures to speed up promotion

2018-11-08 Laurent Rochette <lrochette@electric-cloud.com>
  * 3.1.9.719: Fix #151 saveAllObjects export server properties

2018-09-14 Laurent Rochette <lrochette@electric-cloud.com>
  * 3.1.8.715: Fix #144 Issue promoting on Windows

2018-09-13 Laurent Rochette <lrochette@electric-cloud.com>
  * 3.1.7.712: Fix #140 Change logo file copy to to simple Copy for changeBannerColor

2018-09-07 Laurent Rochette <lrochette@electric-cloud.com>
  * 3.1.6.709: Fix #143 Add recursive option to restoreObjects
  * 3.1.5.706: Fix #135 re-add schedules that did not make the trip to PluginWizard

2018-09-05 Laurent Rochette <lrochette@electric-cloud.com>
  * 3.1.4.702: Fix #72 saveAllObjects now include artifacts and artifact versions

2018-08-30 Laurent Rochette <lrochette@electric-cloud.com>
  * 3.1.3.693:
    Fix #84 Allow banner color change on deploy side as well
    You can now choose your own color
    Fix #136 Logo is now visible also on deploy side

2018-08-16 Laurent Rochette <lrochette@electric-cloud.com>
  * 3.1.2.672: Fix #130 Remove plugin Configuration page and procedures

2018-07-03 Laurent Rochette <lrochette@electric-cloud.com>
  * 3.1.0.629: Convert to PluginWizard/DSL format

2018-06-14 Laurent Rochette <lrochette@electric-cloud.com>
  * 3.0.0.486: Move to Flow 6.0

2018-06-13 Laurent Rochette <lrochette@electric-cloud.com>
  * 3.0.0.485: Move to Flow 6.0

2017-07-38 D. Barringer
  * 2.27.3: Fixed #117: problem with backslash on workspace directory


dsl/procedures/saveAllObjects/steps/saveDeployObjects.pl
#############################################################################
#
#  Script to backup the Deploy objects (application, environment, components,
#     releases, services, ...) in XML or DSL format
#
#  Author: L.Rochette
#
#  Copyright 2015-2019 Electric-Cloud Inc.
#
#     Licensed under the Apache License, Version 2.0 (the "License");
#     you may not use this file except in compliance with the License.
#     You may obtain a copy of the License at
#
#         http://www.apache.org/licenses/LICENSE-2.0
#
#     Unless required by applicable law or agreed to in writing, software
#     distributed under the License is distributed on an "AS IS" BASIS,
#     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#     See the License for the specific language governing permissions and
#     limitations under the License.
#
# History
# ---------------------------------------------------------------------------
# 2019-Feb-11 lrochette Foundation for merge DSL and XML export
# 2019-Feb 21 lrochette Changing paths to match EC-DslDeploy
# 2019-Feb-22 lrochette Fix #166: save catalogs and catalogItems
#                       Fix #165: save dashbaords
# 2019-Apr-16 lrochette Fix format to match EC-DslDeploy
# 2019-Jun-05 lrochette Issue 38: add support for devOpsInsightDataSources
##############################################################################
use File::Path;

$[/myProject/scripts/perlHeaderJSON]

#
# Parameters
#
my $path             = '$[pathname]';
my $pattern          = '$[pattern]';
my $includeACLs      = "$[includeACLs]";
my $includeNotifiers = "$[includeNotifiers]";
my $relocatable      = "$[relocatable]";
my $format           = '$[format]';

#
# Global
#
my $version=getVersion();
my $errorCount  = 0;
my $appCount    = 0;
my $envCount    = 0;
my $compCount   = 0;
my $pipeCount   = 0;
my $relCount    = 0;
my $srcCount    = 0;
my $servCount   = 0;
my $catCount    = 0;
my $itemCount   = 0;
my $dashCount   = 0;
my $widgetCount = 0;
my $filterCount = 0;
my $reportCount = 0;

# Set the timeout to config value or 600 if not set
my $defaultTimeout = getP("/server/EC-Admin/cleanup/config/timeout");
$ec->setTimeout($defaultTimeout? $defaultTimeout : 600);

#
# Get list of Project
my ($success, $xPath) = InvokeCommander("SuppressLog", "getProjects");

# Create the Projects directory
mkpath("$path/projects");
chmod(0777, "$path/projects") or die("Can't change permissions on $path/projects: $!");

foreach my $node ($xPath->findnodes('//project')) {
  my $pName=$node->{'projectName'};
  my $pluginName=$node->{'pluginName',};

  # skip plugins
  next if ($pluginName ne "");

  # skip projects that don't fit the pattern
  next if ($pName !~ /$pattern/$[caseSensitive] ); # / just for the color

  # skip non Default project for version before 6.2
  next if (($pName ne "Default") && ($version < "6.2"));
  printf("Saving Project: %s\n", $pName);

  my $fileProjectName=safeFilename($pName);
  #
  # Save Applications
  #
  mkpath("$path/projects/$fileProjectName/applications");
  chmod(0777, "$path/projects/$fileProjectName/applications");

  my ($success, $xPath) = InvokeCommander("SuppressLog", "getApplications", $pName);
  foreach my $app ($xPath->findnodes('//application')) {
    my $appName=$app->{'applicationName'};

    # skip applications that don't fit the pattern
    next if (($pName  eq "Default") && ($appName !~ /$pattern/$[caseSensitive] )); # / just for the color

    my $fileAppName=safeFilename($appName);
    printf("  Saving Application: %s\n", $appName);

    mkpath("$path/projects/$fileProjectName/applications/$fileAppName");
    chmod(0777, "$path/projects/$fileProjectName/applications/$fileAppName");
    my ($success, $res, $errMsg, $errCode) =
      backupObject($format,
        "$path/projects/$fileProjectName/applications/$fileAppName/application",
        "/projects[$pName]applications[$appName]",
        $relocatable, $includeACLs, $includeNotifiers);

    if (! $success) {
      printf("  Error exporting application %s", $appName);
      printf("  %s: %s\n", $errCode, $errMsg);
      $errorCount++;
    }
    else {
      $appCount++;
    }

    #
    # backup Components
    mkpath("$path/projects/$fileProjectName/applications/$fileAppName/components");
    chmod(0777, "$path/projects/$fileProjectName/applications/$fileAppName/components");

    my ($ok, $json) = InvokeCommander("SuppressLog", "getComponents", $pName, {applicationName => $appName});
    foreach my $comp ($json->findnodes("//component")) {
      my $compName=$comp->{'componentName'};
      my $fileCompName=safeFilename($compName);
      printf("    Saving Component: %s\n", $compName);

      my ($success, $res, $errMsg, $errCode) =
        backupObject($format,
          "$path/projects/$fileProjectName/applications/$fileAppName/components/$fileCompName",
          "/projects[$pName]applications[$appName]components[$compName]",
          $relocatable, $includeACLs, $includeNotifiers);

      if (! $success) {
        printf("  Error exporting component %s in application", $compName, $appName);
        printf("  %s: %s\n", $errCode, $errMsg);
        $errorCount++;
      }
      else {
        $compCount++;
      }

    }
  }

  #
  # Save Environments definitions
  #
  mkpath("$path/projects/$fileProjectName/environments");
  chmod(0777, "$path/projects/$fileProjectName/environments");

  my ($success, $xPath) = InvokeCommander("SuppressLog", "getEnvironments", $pName);
  foreach my $proc ($xPath->findnodes('//environment')) {
    my $envName=$proc->{'environmentName'};

    # skip environments that don't fit the pattern
    next if (($pName  eq "Default") && ($envName !~ /$pattern/$[caseSensitive] ));  # / just for the color

    my $fileEnvName=safeFilename($envName);
    printf("  Saving Environment: %s\n", $envName);

    my ($success, $res, $errMsg, $errCode) =
      backupObject($format, "$path/projects/$fileProjectName/environments/$fileEnvName",
        "/projects[$pName]environments[$envName]",
        $relocatable, $includeACLs, $includeNotifiers);

    if (! $success) {
      printf("  Error exporting environment %s", $envName);
      printf("  %s: %s\n", $errCode, $errMsg);
      $errorCount++;
    }
    else {
      $envCount++;
    }
  }

  #
  # Export pipelines if the version is recent enough
  #
  if (compareVersion($version, "6.0") < 0) {
    printf("WARNING: Version 6.0 or greater is required to save Pipeline objects");
  } else {
    # Save pipeline definitions
    #
    mkpath("$path/projects/$fileProjectName/pipelines");
    chmod(0777, "$path/projects/$fileProjectName/pipelines");

    my ($success, $xPath) = InvokeCommander("SuppressLog", "getPipelines", $pName);
    foreach my $proc ($xPath->findnodes('//pipeline')) {
      my $pipeName=$proc->{'pipelineName'};

      # skip pipelines that don't fit the pattern
      next if (($pName  eq "Default") && ($pipeName !~ /$pattern/$[caseSensitive] )); # / just for the color

      my $filePipeName=safeFilename($pipeName);
      printf("  Saving Pipeline: %s\n", $pipeName);

      my ($success, $res, $errMsg, $errCode) =
        backupObject($format,
          "$path/projects/$fileProjectName/pipelines/$filePipeName",
          "/projects[$pName]pipelines[$pipeName]",
          $relocatable, $includeACLs, $includeNotifiers);

      if (! $success) {
        printf("  Error exporting pipeline %s", $pipeName);
        printf("  %s: %s\n", $errCode, $errMsg);
        $errorCount++;
      }
      else {
        $pipeCount++;
      }
    }         # Pipeline loop
  }           # Version greater than 6.0


  # Export releases if the version is recent enough
  #
  if (compareVersion($version, "6.1") < 0) {
    printf("WARNING: Version 6.1 or greater is required to save Release objects");
  } else {
    # Save release definitions
    #
    mkpath("$path/projects/$fileProjectName/releases");
    chmod(0777, "$path/projects/$fileProjectName/releases");

    my ($success, $xPath) = InvokeCommander("SuppressLog", "getReleases", $pName);
    foreach my $proc ($xPath->findnodes('//release')) {
      my $relName=$proc->{'releaseName'};

      # skip releases that don't fit the pattern
      next if (($pName  eq "Default") && ($relName !~ /$pattern/$[caseSensitive] ));  # / just for the color

      printf("  Saving Release: %s\n", $relName);
      my $fileRelName=safeFilename($relName);
      mkpath("$path/projects/$fileProjectName/releases/$fileRelName");
      chmod(0777, "$path/projects/$fileProjectName/releases/$fileRelName");

      my ($success, $res, $errMsg, $errCode) =
        backupObject($format,
          "$path/projects/$fileProjectName/releases/$fileRelName/release",
          "/projects[$pName]releases[$relName]",
          $relocatable, $includeACLs, $includeNotifiers);

      if (! $success) {
        printf("  Error exporting release %s", $relName);
        printf("  %s: %s\n", $errCode, $errMsg);
        $errorCount++;
      }
      else {
        $relCount++;
      }

      if (compareVersion($version, "9.0") < 0) {
        printf("WARNING: Version 9.0 or greater is required to save DOIS sources");
      } else {
        mkpath("$path/projects/$fileProjectName/releases/$fileRelName/devOpsInsightDataSources");
        chmod(0777, "$path/projects/$fileProjectName/releases/$fileRelName/devOpsInsightDataSources");
        my ($suc2, $xPath2) = InvokeCommander("SuppressLog", "getDevOpsInsightDataSources", $pName, $relName);
        foreach my $src ($xPath2->findnodes('//devOpsInsightDataSource')) {
          my $srcName=$src->{'devOpsInsightDataSourceName'};
          printf("    Saving devOpsInsightDataSource: %s\n", $srcName);
          my $fileSrcName=safeFilename($srcName);
          mkpath("$path/projects/$fileProjectName/releases/$fileRelName/devOpsInsightDataSources/$fileSrcName");
          chmod(0777, "$path/projects/$fileProjectName/releases/$fileRelName/devOpsInsightDataSources/$fileSrcName");
          ($success, $res, $errMsg, $errCode) =
            backupObject($format,
              "$path/projects/$fileProjectName/releases/$fileRelName/devOpsInsightDataSources/$fileSrcName/devOpsInsightDataSource",
              "/projects[$pName]releases[$relName]devOpsInsightDataSources[$srcName]",
              $relocatable, $includeACLs, $includeNotifiers);

          if (! $success) {
            printf("    Error exporting devOpsInsightDataSource %s", $srcName);
            printf("    %s: %s\n", $errCode, $errMsg);
            $errorCount++;
          }
          else {
            $srcCount++;
          }
        }
      }
    }         # Release loop
  }           # Version greater than 6.1

  # Export services if the version is recent enough
  #
  if (compareVersion($version, "8.1") < 0) {
    printf("WARNING: Version 8.1 or greater is required to save Services objects");
  } else {
    # Save release definitions
    #
    mkpath("$path/projects/$fileProjectName/services");
    chmod(0777, "$path/projects/$fileProjectName/services");

    my ($success, $xPath) = InvokeCommander("SuppressLog", "getServices", $pName);
    foreach my $proc ($xPath->findnodes('//service')) {
      my $servName=$proc->{'serviceName'};

      # skip services that don't fit the pattern
      next if (($pName  eq "Default") && ($servName !~ /$pattern/$[caseSensitive] ));  # / just for the color

      my $fileServName=safeFilename($servName);
      printf("  Saving Service: %s\n", $servName);
      mkpath("$path/projects/$fileProjectName/services/$fileServName");
      chmod(0777, "$path/projects/$fileProjectName/services/$fileServName");

      my ($success, $res, $errMsg, $errCode) =
        backupObject($format,
          "$path/projects/$fileProjectName/services/$fileServName/service",
          "/projects[$pName]services[$servName]",
          $relocatable, $includeACLs, $includeNotifiers);

      if (! $success) {
        printf("  Error exporting service %s", $servName);
        printf("  %s: %s\n", $errCode, $errMsg);
        $errorCount++;
      }
      else {
        $servCount++;
      }
    }         # service loop
  }           # Version greater than 8.1

  # Export catalogs if the version is recent enough
  #
  if (compareVersion($version, "7.3") < 0) {
    printf("WARNING: Version 7.3 or greater is required to save Catalog objects");
  } else {
    # Save catalog definitions
    #
    mkpath("$path/projects/$fileProjectName/catalogs");
    chmod(0777, "$path/projects/$fileProjectName/catalogs");

    my ($success, $xPath) = InvokeCommander("SuppressLog", "getCatalogs", $pName);
    foreach my $proc ($xPath->findnodes('//catalog')) {
      my $catName=$proc->{'catalogName'};

      # skip catalogs that don't fit the pattern
      next if (($pName  eq "Default") && ($catName !~ /$pattern/$[caseSensitive] ));  # / just for the color

      my $fileCatName=safeFilename($catName);
      printf("  Saving Catalog: %s\n", $catName);

     mkpath("$path/projects/$fileProjectName/catalogs/$fileCatName");
     chmod(0777, "$path/projects/$fileProjectName/catalogs/$fileCatName");

      my ($success, $res, $errMsg, $errCode) =
        backupObject($format,
          "$path/projects/$fileProjectName/catalogs/$fileCatName/catalog",
          "/projects[$pName]catalogs[$catName]",
          $relocatable, $includeACLs, $includeNotifiers);

      if (! $success) {
        printf("  Error exporting catalog %s", $catName);
        printf("  %s: %s\n", $errCode, $errMsg);
        $errorCount++;
      }
      else {
        $servCount++;
      }

      # backup catalogItems
      mkpath("$path/projects/$fileProjectName/catalogs/$fileCatName/catalogItems");
      chmod(0777, "$path/projects/$fileProjectName/catalogs/$fileCatName/catalogItems");
      my ($ok, $json) = InvokeCommander("SuppressLog", "getCatalogItems", $pName, $catName);
      foreach my $item ($json->findnodes("//catalogItem")) {
        my $itemName=$item->{'catalogItemName'};
        my $fileItemName=safeFilename($itemName);
        printf("    Saving Catalog Item: %s\n", $itemName);
        mkpath("$path/projects/$fileProjectName/catalogs/$fileCatName/catalogItems/$fileItemName");
        chmod(0777, "$path/projects/$fileProjectName/catalogs/$fileCatName/catalogItems/$fileItemName");

        my ($success, $res, $errMsg, $errCode) =
          backupObject($format,
            "$path/projects/$fileProjectName/catalogs/$fileCatName/catalogItems/$fileItemName/catalogItem",
            "/projects[$pName]catalogs[$catName]catalogItems[$itemName]",
            $relocatable, $includeACLs, $includeNotifiers);

        if (! $success) {
          printf("  Error exporting catalog item %s in catalog", $itemName, $catName);
          printf("  %s: %s\n", $errCode, $errMsg);
          $errorCount++;
        }
        else {
          $itemCount++;
        }

      }

    }         # catalog loop
  }           # Version greater than 7.3


  # Export dashboards if the version is recent enough
  #
  if (compareVersion($version, "7.1") < 0) {
    printf("WARNING: Version 7.1 or greater is required to save Dashboards objects");
  } else {
    # Save dashboard definitions
    #
    mkpath("$path/projects/$fileProjectName/dashboards");
    chmod(0777, "$path/projects/$fileProjectName/dashboards");

    my ($success, $xPath) = InvokeCommander("SuppressLog", "getDashboards", {projectName => $pName});
    foreach my $dash ($xPath->findnodes('//dashboard')) {
      my $dashName=$dash->{'dashboardName'};

      # skip dashboards that don't fit the pattern
      next if (($pName  eq "Default") && ($dashName !~ /$pattern/$[caseSensitive] ));  # / just for the color

      my $fileDashName=safeFilename($dashName);
      printf("  Saving Dashboard: %s\n", $dashName);
      mkpath("$path/projects/$fileProjectName/dashboards/$fileDashName");
      chmod(0777, "$path/projects/$fileProjectName/dashboards/$fileDashName");

      my ($success, $res, $errMsg, $errCode) =
        backupObject($format,
          "$path/projects/$fileProjectName/dashboards/$fileDashName/dashboard",
          "/projects[$pName]dashboards[$dashName]",
          $relocatable, $includeACLs, $includeNotifiers);

      if (! $success) {
        printf("  Error exporting dashboard %s", $dashName);
        printf("  %s: %s\n", $errCode, $errMsg);
        $errorCount++;
      }
      else {
        $dashCount++;
      }

      # backup widget
      my ($ok, $json) = InvokeCommander("SuppressLog", "getWidgets", $pName, $dashName);
      mkpath("$path/projects/$fileProjectName/dashboards/$fileDashName/widgets");
      chmod(0777, "$path/projects/$fileProjectName/dashboards/$fileDashName/widgets");
      foreach my $widget ($json->findnodes("//widget")) {
        my $widgetName=$widget->{'widgetName'};
        my $fileWidgetName=safeFilename($widgetName);
        printf("    Saving Widget: %s\n", $widgetName);
        mkpath("$path/projects/$fileProjectName/dashboards/$fileDashName/widgets/$fileWidgetName");
        chmod(0777, "$path/projects/$fileProjectName/dashboards/$fileDashName/widgets/$fileWidgetName");

        my ($success, $res, $errMsg, $errCode) =
          backupObject($format,
            "$path/projects/$fileProjectName/dashboards/$fileDashName/widgets/$fileWidgetName/widget",
            "/projects[$pName]dashboards[$dashName]widgets[$widgetName]",
            $relocatable, $includeACLs, $includeNotifiers);

        if (! $success) {
          printf("  Error exporting widget %s in dashboard", $widgetName, $dashName);
          printf("  %s: %s\n", $errCode, $errMsg);
          $errorCount++;
        }
        else {
          $widgetCount++;
        }

      }       # widget Loop

      # backup reportingFilters
      mkpath("$path/projects/$fileProjectName/dashboards/$fileDashName/reportingFilters");
      chmod(0777, "$path/projects/$fileProjectName/dashboards/$fileDashName/reportingFilters");
      my ($ok, $json) = InvokeCommander("SuppressLog", "getReportingFilters", $pName, $dashName);
      foreach my $filter ($json->findnodes("//reportingFilter")) {
        my $filterName=$filter->{'reportingFilterName'};
        my $fileFilterName=safeFilename($filterName);
        printf("    Saving ReportingFilter: %s\n", $filterName);
        mkpath("$path/projects/$fileProjectName/dashboards/$fileDashName/reportingFilters/$fileFilterName");
        chmod(0777, "$path/projects/$fileProjectName/dashboards/$fileDashName/reportingFilters/$fileFilterName");

        my ($success, $res, $errMsg, $errCode) =
          backupObject($format,
            "$path/projects/$fileProjectName/dashboards/$fileDashName/reportingFilters/$fileFilterName/reportingFilter",
            "/projects[$pName]dashboards[$dashName]reportingFilters[$filterName]",
            $relocatable, $includeACLs, $includeNotifiers);

        if (! $success) {
          printf("  Error exporting filter %s in dashboard", $filterName, $dashName);
          printf("  %s: %s\n", $errCode, $errMsg);
          $errorCount++;
        }
        else {
          $filterCount++;
        }

      }       # Filter Loop

    }         # dashboard loop

    # Save reports
    #
    mkpath("$path/projects/$fileProjectName/reports");
    chmod(0777, "$path/projects/$fileProjectName/reports");
    my ($success, $xPath) = InvokeCommander("SuppressLog", "getReports", {projectName => $pName});
    foreach my $report ($xPath->findnodes('//report')) {
      my $reportName=$report->{'reportName'};

      # skip reports that don't fit the pattern
      next if (($pName  eq "Default") && ($reportName !~ /$pattern/$[caseSensitive] ));  # / just for the color

      my $fileReportName=safeFilename($reportName);
      printf("  Saving Report: %s\n", $reportName);
      mkpath("$path/projects/$fileProjectName/reports/$fileReportName");
      chmod(0777, "$path/projects/$fileProjectName/reports/$fileReportName");

       my ($success, $res, $errMsg, $errCode) =
         backupObject($format,
          "$path/projects/$fileProjectName/reports/$fileReportName/report",
          "/projects[$pName]reports[$reportName]",
          $relocatable, $includeACLs, $includeNotifiers);

      if (! $success) {
        printf("  Error exporting report %s", $reportName);
        printf("  %s: %s\n", $errCode, $errMsg);
        $errorCount++;
      }
      else {
        $reportCount++;
      }
    }         # report loop
  }           # Version greater than 7.3

}             # project loop

my $str="";
$str .= createExportString($appCount, "application")         if ($appCount);
$str .= createExportString($envCount, "environment")         if ($envCount);
$str .= createExportString($compCount, "component")          if ($compCount);
$str .= createExportString($pipeCount, "pipeline")           if ($pipeCount);
$str .= createExportString($relCount, "release")             if ($relCount);
$str .= createExportString($srcCount, "source")              if ($srcCount);
$str .= createExportString($servCount, "service")            if ($servCount);
$str .= createExportString($catCount, "catalog")             if ($catCount);
$str .= createExportString($itemCount, "catalog item")       if ($itemCount);
$str .= createExportString($reportCount, "report")           if ($reportCount);
$str .= createExportString($dashCount, "dashboard")          if ($dashCount);
$str .= createExportString($widgetCount, "widget")           if ($widgetCount);
$str .= createExportString($filterCount, "reporting filter") if ($filterCount);

$ec->setProperty("preSummary", $str);

$ec->setProperty("/myJob/papplicationExported",    $appCount);
$ec->setProperty("/myJob/environmentExported",     $envCount);
$ec->setProperty("/myJob/componentExported",       $compCount);
$ec->setProperty("/myJob/pipelineExported",        $pipeCount);
$ec->setProperty("/myJob/releaseExported",         $relCount);
$ec->setProperty("/myJob/sourceExported",          $srcCount);
$ec->setProperty("/myJob/serviceExported",         $servCount);
$ec->setProperty("/myJob/catalogExported",         $catCount);
$ec->setProperty("/myJob/catalogItemExported",     $itemCount);
$ec->setProperty("/myJob/dashboardExported",       $dashCount);
$ec->setProperty("/myJob/widgetExported",          $widgetCount);
$ec->setProperty("/myJob/reportingFilterExported", $filterCount);
$ec->setProperty("/myJob/reportExported",          $reportCount);
exit($errorCount);

$[/myProject/scripts/perlBackupLib]
$[/myProject/scripts/perlLibJSON]
